x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/persistence/TrainedModelProvider.java

```diff
@@ -51,83 +55,178 @@
-                                    RestStatus.INTERNAL_SERVER_ERROR,
-                                    e,
-                                    trainedModelConfig.getModelId()));
-                        }
-                    }));
-        } catch (IOException e) {
-            // not expected to happen but for the sake of completeness
-            listener.onFailure(new ElasticsearchParseException(
-                Messages.getMessage(Messages.INFERENCE_FAILED_TO_SERIALIZE_MODEL, trainedModelConfig.getModelId()),
-                e));
-        }
+        ActionListener<IndexResponse> putDefinitionListener = ActionListener.wrap(
+            r -> listener.onResponse(true),
+            e -> {
+                logger.error(new ParameterizedMessage(
+                    "[{}] failed to store trained model definition for inference", trainedModelConfig.getModelId()), e);
+                if (ExceptionsHelper.unwrapCause(e) instanceof VersionConflictEngineException) {
+                    listener.onFailure(new ResourceAlreadyExistsException(
+                        Messages.getMessage(Messages.INFERENCE_TRAINED_MODEL_EXISTS, trainedModelConfig.getModelId())));
+                } else {
+                    listener.onFailure(
+                        new ElasticsearchStatusException(Messages.INFERENCE_FAILED_TO_STORE_MODEL,
+                            RestStatus.INTERNAL_SERVER_ERROR,
+                            e,
+                            trainedModelConfig.getModelId()));
+                }
+            }
+        );
+
+        ActionListener<IndexResponse> putConfigListener = ActionListener.wrap(
+            r -> {
+                if (trainedModelConfig.getDefinition() != null) {
+                    indexObject(TrainedModelDefinition.docId(trainedModelConfig.getModelId()),
+                        trainedModelConfig.getDefinition(),
+                        putDefinitionListener);
+                } else {
+                    listener.onResponse(true);
+                }
+            },
+            e -> {
+                logger.error(new ParameterizedMessage(
+                    "[{}] failed to store trained model for inference", trainedModelConfig.getModelId()), e);
+                if (ExceptionsHelper.unwrapCause(e) instanceof VersionConflictEngineException) {
+                    listener.onFailure(new ResourceAlreadyExistsException(
+                        Messages.getMessage(Messages.INFERENCE_TRAINED_MODEL_EXISTS, trainedModelConfig.getModelId())));
+                } else {
+                    listener.onFailure(
+                        new ElasticsearchStatusException(Messages.INFERENCE_FAILED_TO_STORE_MODEL,
+                            RestStatus.INTERNAL_SERVER_ERROR,
+                            e,
+                            trainedModelConfig.getModelId()));
+                }
+            }
+        );
+
+        indexObject(trainedModelConfig.getModelId(), trainedModelConfig, putConfigListener);
```

Comment:
I think we should consider indexing both objects in a bulk request. That way we'd be refreshing the index once and we can also handle errors in one place.